Refactor Modulation Extension Namespace

Atheatos opened the pull request:
Refactor the modulation extension based on discussions from https://github.com/gnuradio/SigMF/pull/87:
 - rename the namespace to `signal` from `modulation`
 - add a free-form `label` field for a short-hand description of the signal under `signal:emission`
 - move the signal descriptor fields to an object `Detail`, nested under `signal:emission`
 - add `channel` to `Detail` so that the communication system's bandwidth channel can be notated
 - rename `system` to `standard` under `Detail`
 - add an object `Emitter`, nested under `signal:emission`, to describe the hardware driving the communication system
 - update examples

I also added a `class_variant` field to `Detail` to cover less straightforward modulations such as pi/4-DQPSK and GMSK. This wasn't discussed in the other PR or previous issues, but something like it is necessary for the signals that I have. I'm not sure it's the best approach since its contents are open to interpretation, however. For example, should its value be `G` or `gaussian` for GMSK? How about `pi/4` vs `π/4`? Maybe a separate variant field for each class, perhaps in a completely different extension?

Here's how it looks for pi/4-DQPSK in the example I added in the spec:
```json
"detail": {
    "type": "digital",
    "class": "psk",
    "order": 4,
    "symbol_variant": "differential",
    "class_variant": "pi/4"
}
```

Teque5 commented:
Nice Job. This looks pretty well thought-out.

bhilburn commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+
+|name|required|type|description|
+|----|--------------|-------|----------|
+|`label`|false|string|e.g., wifi|
```
I know this field seems relatively straight-forward, but it would be good to specifically call out it's meant to be user-defined shortform. Maybe something like:

```txt
A high-level shortform description useful to human readers (e.g., "wifi", "LTE")
```

bhilburn commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+|----|--------|----|-----------|
+|`seid`|false|uint|Unique ID of the emitter|
+|`manufacturer`|false|string|Manufacterer of the hardware used to emit the signal|
+|`latitude`|false|float|Latitude|
+|`longitude`|false|float|longitude|
```
We should probably follow the NTIA's example, here, and specify the units as "decimal degrees" for both long & lat. Near the top of @MichaelCotton's comment, [here](https://github.com/gnuradio/SigMF/pull/87#issuecomment-508749532)

bhilburn reviewed (commented):
Excellent job, @Atheatos, taking all of the discussion and ideas over nearly 3 years of PR hacking and making a useful PR out of it! Thanks so much!

I think with a couple of minor fixes (see comments), this is ready for merge 👍

Atheatos commented:
Happy to hear that it looks good; thanks for the reviews.

I don't have anything to add to your description of the `label` field, so I just added it verbatim. Since adding a "units" column wouldn't really make sense for that table, I just added it to the description.

bhilburn commented:
Would like to get review from @jacobagilbert / @n-west, then want to merge ASAP.

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+
+|name|required|type|description|
+|----|--------------|-------|----------|
+|`label`|false|string|A high-level shortform description useful to human readers (e.g., "wifi", "LTE")|
```
I believe the `label` field is extremely common and should actually be moved to core (we could also call this `description` to be consistent with `global`), but this is really something I think should be up a level and more generic.

bhilburn replied:
This is an interesting point, @jacobagilbert. Currently, that's meant to be filled by the `comment` field, but I think what you're really highlighting is that it would be worthwhile to have a short-form human **and machine** -readable field that's in `core`. 

For purposes of process, I recommend we leave this field here, and then file a new issue & PR discussing and making that change across the codebase to keep the change atomic.

jacobagilbert replied:
Yes, thats right. 

Are you suggesting that we merge this here then separately move it to core after this is merged?

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+|name|required|type|description|
+|----|--------------|-------|----------|
+|`label`|false|string|A high-level shortform description useful to human readers (e.g., "wifi", "LTE")|
+|`detail`|false|[Detail](signal.sigmf-ext.md#the-detail-object)|Emission details (standard, modulation, etc.)|
```
I think `signal:emitter` should be directly within the annotations object.

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+|----|--------------|-------|----------|
+|`label`|false|string|A high-level shortform description useful to human readers (e.g., "wifi", "LTE")|
+|`detail`|false|[Detail](signal.sigmf-ext.md#the-detail-object)|Emission details (standard, modulation, etc.)|
+|`emitter`|false|[Emitter](signal.sigmf-ext.md#the-emitter-object)|Emitter details (manufacturer, geo coordinates, etc.)|
```
Same as above, `signal:emission_detail` or `signal:emission` should be directly within the annotations object. This also removes one of the hierarchical levels making this a bit simpler.

bhilburn replied:
Pulling from the examples section, this is what we have currently:

```json
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "signal:emission": {
            "label": "wifi",
            "detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
            },
            "emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
            }
        }
    }]
}
```

What you are suggesting is change it to what would look like:

```json
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "signal:emission": {
            "label": "wifi"
        },
        "signal:detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
         },
         "signal:emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
        }
    }]
}
```

I agree with you that it simplifies the "hierarchy" of reaching into the `emission`, but it also bloats what sits at the level of the hierarchy for the annotation, and from a semantic perspective, I actually think it makes less sense?

Not sold one way or the other. Thoughts from @n-west, @Teque5, @Atheatos?

jacobagilbert replied:
This is what I am specifically proposing:

```JSON
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "core:label": "wifi",
        "signal:detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
         },
         "signal:emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
        }
    }]
}

largely because in my use of SigMF annotations there are many times i would *only* have `label` or `signal:emitter` or `signal:detail` information to include and the extra layer of hierarchy. If we move `label` to `core` as i think we agree is a good idea, this is two functionally separate concepts (details about the emitter, and details about the emission).

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+|`seid`|false|uint|Unique ID of the emitter|
+|`manufacturer`|false|string|Manufacterer of the hardware used to emit the signal|
+|`latitude`|false|float|Latitude (decimal degrees)|
+|`longitude`|false|float|Longitude (decimal degrees)|
```
Not to delay merging, but I think we should come to a consensus on how we handle geospatial metadata before we merge this. If we are going to use GeoJSON we should just put a `location` field here and allow the same format.

If we end up sticking with discrete coordinate fields, i suggest we add `altitude` and define the format.

See #73 and #84

bhilburn replied:
Agreed, this should be made consistent with the outcomes of that discussion. As with the `label` discussion above, I'd like to /not/ hold this PR on this topic, and then once that decision is made, make that change across the code base.

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+|`pm`|(analog) phase modulation|
+|`ssb`|(analog) single side-band|
+|`dsb`|(analog) dual side-band|
+|`vsb`|(analog) vestigal side-band|
```
This is minor, but VSB can also used with digital modulations (see: ATSC)

bhilburn replied:
Interesting, I didn't know that! @Atheatos, want to make this `(analog / digital)` for completeness?

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+
+###### The `bandwidth` Field
+
+The `bandwidth` field has a float value that describes the channel bandwidth of the signal. Note that this is different from what is reported in the `core` namespace within an annotation, describing the occupied spectrum of a signal, which may or may not be comparable to the actual channel bandwidth of the communications system.
```
Can we change this name to be more specific then? `channel_bw` or `channel_bandwidth`?

bhilburn replied:
Good point, and agreed with @jacobagilbert.

jacobagilbert commented on extensions/signal.sigmf-ext.md:
```diff
@@ -0,0 +1,285 @@
+
+The `class_variant` field describes any modifier to the modulation class not covered by any of the other fields. Examples include pi/4-DQPSK and GMSK.
+
+## Examples
```
Awesome providing examples!!

jacobagilbert reviewed (changes requested):
I really like this overall, very well thought out.

I do think that there is some overlap between this and some other objects and that we might be able to flatten this by one level before merging; comments are inline.

Teque5 commented:
Suggestion for extensions in general: consider section numbering similar to [how NTIA has adopted sigmf extensions](https://github.com/NTIA/sigmf-ns-ntia/blob/master/ntia-emitter.sigmf-ext.md).

bhilburn reviewed (changes requested):
Responded to review comments from @jacobagilbert. I think there are two small changes to make, one outstanding discussion to close (see hierarchy of `emission`), and two other items that will be addressed outside of this PR.

jacobagilbert commented:
@Atheatos Do you have time to work on this? If not, are you OK if I hijack this to get the changes made for inclusion in v1.0?

bhilburn commented:
@jacobagilbert - I think you're clear to pick this up and run with it. Please do it on top of @Atheatos's commits so his contributions remain represented in-tree.

Atheatos commented:
@jacobagilbert @bhilburn 
Hey, sorry for the delay in responding. Indeed, I'm not able to work on this right now, so I appreciate you moving forward to get the change integrated for the v.1.0 release

jacobagilbert commented:
Thanks @Atheatos I just opened a new PR though. It has your commit history so your contribution will be retained.
